dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -231,38 +232,29 @@ Block DMFileReader::readWithFilter(const IColumn::Filter & filter)
         //      When i = next_pack_id + 5, call read() to read {next_pack_id + 3, next_pack_id + 4, next_pack_id + 5}th packs
         if (pack_res[pack_id].isUse() && (pack_id + 1 == pack_res.size() || !pack_res[pack_id + 1].isUse()))
         {
-            Block block = read();
-            size_t rows = block.rows();
-
-            if (size_t passed_count = countBytesInFilter(filter, offset, rows); passed_count != rows)
+            read_rows += pack_stats[pack_id].rows;
+            IColumn::Filter block_filter(filter.cbegin() + offset, filter.cbegin() + offset + read_rows);
+            Block block = read(&block_filter);
+            size_t passed_count = countBytesInFilter(block_filter);
+            for (size_t i = 0; i < block.columns(); ++i)
             {
-                std::vector<size_t> positions;
-                positions.reserve(passed_count);
-                for (size_t p = offset; p < offset + rows; ++p)
-                {
-                    if (filter[p])
-                        positions.push_back(p - offset);
-                }
-                for (size_t i = 0; i < block.columns(); ++i)
-                {
-                    columns[i]->insertDisjunctFrom(*block.getByPosition(i).column, positions);
-                }
+                auto col = block.getByPosition(i).column;
+                // Some columns may only deserialize the passed rows.
+                if (col->size() != passed_count)
+                    col = col->filter(block_filter, passed_count);
```

Comment:
We'd better ensure all the columns return by `read(IColumn::Filter * filter)` has the same number of rows. But not handle it in this for-loop.

Reply:
I have addressed in https://github.com/pingcap/tiflash/pull/9687, since we will rewrite this soon, so just keep it in this PR.